Add a backport of generic `NamedTuple`s (#44)
python/typing_extensions: README.md

  introspectable at runtime. In order to access overloads with
  `typing_extensions.get_overloads()`, you must use
  `@typing_extensions.overload`.
- `NamedTuple` was changed in Python 3.11 to allow for multiple inheritance

>   with `typing.Generic` The implementation of `NamedTuple` was also changed in

Reviewer: ```suggestion
  with `typing.Generic`. The implementation of `NamedTuple` was also changed in
```

> - `NamedTuple` was changed in Python 3.11 to allow for multiple inheritance
>   with `typing.Generic` The implementation of `NamedTuple` was also changed in
>   3.9 so that `NamedTuple` became a function rather than a class.

Reviewer 2: Maybe switch it around so that the earlier change is earlier in the text?

```suggestion
- The implementation of `NamedTuple` was changed in Python 3.9 so that `NamedTuple` became
  a function rather than a class. Since Python 3.11 `NamedTuple` allows for multiple inheritance
  with `typing.Generic` 
```

(Although the 3.11 change is of course the more relevant one.)

Author: Yeah, the 3.9 change is basically an implementation detail that's only being backported to keep the `typing_extensions` code simple (I didn't want to have to introduce a whole new code branch so that `typing_extensions.NamedTuple` would match `typing.NamedTuple` and be a class in 3.8). I'd rather lead with the 3.11 change, I think.

>   3.9 so that `NamedTuple` became a function rather than a class.

Reviewer: Why do we need to mention how it's implemented? How does it affect users?

Author: I included this because I thought it might be confusing for users on 3.8 if they realised that `typing.NamedTuple` was a class, but `typing_extensions.NamedTuple` was a function. (It _is_ a bit weird.) But I'm happy to delete it if you don't think that's an issue.

Reviewer: I'd rather remove it, users shouldn't care about implementation details.
